### PR TITLE
core/sched: add sched_change_priority()

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

This PR is a rebase of the `sched_change_priority()` function from https://github.com/RIOT-OS/RIOT/pull/7445 (introduced in commit https://github.com/RIOT-OS/RIOT/pull/7445/commits/976d17f8b1076568c0df0e6053893557f441ce2b) with some changes:

1. No module required to unlock this
    - The linker does a fine job at garbage collecting the function, if it is unused
2. The function works as expected in corner cases
    - If the modified thread is the running one, this will work
    - If the modified thread is pending and becomes a higher priority than the active one, it will now preempt the running one

### Testing procedure

A test application was added that contains an automatic test. E.g.

```
$ make BOARD=nucleo-f767zi -C tests/sched_change_priority flash test
[...]
socat - open:/dev/ttyACM0,b115200,echo=0,raw 
main(): This is RIOT! (Version: 2022.01-devel-305-ga9249-core/change_prio)
[t3] Setting my priority to THREAD_PRIORITY_MAIN + 2 = 9
[main] Use shell command "nice" to increase prio of t3.
[main] If it works, it will run again. The "hint" cmd can be useful.
> main(): This is RIOT! (Version: 2022.01-devel-305-ga9249-core/change_prio)
[t3] Setting my priority to THREAD_PRIORITY_MAIN + 2 = 9
[main] Use shell command "nice" to increase prio of t3.
[main] If it works, it will run again. The "hint" cmd can be useful.
> hint

> 
> hint
Run "nice 3 6"
> nice 3 6
nice 3 6
[t3] Running again.
make: Leaving directory '/home/maribu/Repos/software/RIOT/tests/sched_change_priority'

$ echo $?
0
```

### Issues/PRs references

https://github.com/RIOT-OS/RIOT/pull/7445

### CODE REVIEW

File: core/sched.c [hunk 1]
```diff
@@ -302,3 +316,36 @@ void sched_register_cb(void (*callback)(kernel_pid_t, kernel_pid_t))
...
+void sched_change_priority(thread_t *thread, uint8_t priority)
+{
+    assert(thread && (priority < SCHED_PRIO_LEVELS));
+
+    if (thread->priority == priority) {
+        return;
+    }
+
+    if (thread_is_active(thread)) {
+        _runqueue_pop(thread);
```
Comment (kaspar030) [inline core/sched.c:329]: this needs some interrupt disabling.

**Comment by kaspar030** [comment]
LGTM, let's get some experience for this.
Please uncrustify and squash!

**Comment by fjmolinas** [comment]
This might need a rebase to pass static checks when https://github.com/RIOT-OS/RIOT/pull/17104 is in...

**Comment by maribu** [comment]
> This might need a rebase to pass static checks when #17104 is in...

Did so just in case. Also added a `Makefile.ci` and squashed it right in.

**Review by benpicco — APPROVED** [review]
If @kaspar030 is OK with it on a conceptual level, the change looks goodⁿᶦᶜᵉ to me.

**Comment by kaspar030** [comment]
There's a slight performance drop (bench_thread_flags_pingpong on nrf52840dk):

```
main(): This is RIOT! (Version: 2022.01-devel-330-gea1f4-HEAD)
main starting
{ "result" : 112675, "ticks" : 568 }
Help: Press s to start test, r to print it is ready
START
main(): This is RIOT! (Version: 2022.01-devel-333-g034bdf-pr17093)
main starting
{ "result" : 111304, "ticks" : 575 }
```
(HEAD is the master fork point).

It's  ~1%. So room for getting some of that back in the future. :)

**Review by kaspar030 — APPROVED** [review]
ACK.

**Comment by kaspar030** [comment]
Let me suggest a rebase to get a current CI run.

**Comment by kfessel** [comment]
there seems to be a slight problem with that Python f-strings (introduced in Python 3.6)

The only test user in master of that python feature in master is the test gnrc_dhcpv6_client, but that is a test with config and therefore not run by murdock

File: tests/sched_change_priority/tests/01-run.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import sys
+from testrunner import run
+
+
+def testfunc(child):
+    child.sendline("hint")
+    child.expect(r"Run \"nice (\d+) (\d+)\"\r\n")
+    pid = int(child.match.group(1))
+    prio = int(child.match.group(2))
+    child.sendline(f"nice {pid} {prio}")
```
Comment (fjmolinas) [inline tests/sched_change_priority/tests/01-run.py:12]:
```suggestion
    child.sendline("nice {} {}".format(pid, prio))
```

**Review by fjmolinas — COMMENTED** [review]
Let's not stall this one because the python version in rpi is old...

**Comment by maribu** [comment]
I squashed the suggested fix right in to save a CI cycle.

**Comment by benpicco** [comment]
Only the usual `run_test/tests/pkg_mbedtls/samr21-xpro:gnu` is failing.

**Pull request merged**